### DISCUSSION TITLE
How to retrieve the index of vertex in CGAL_3D_MESH

### DISCUSSION THREAD

**Discussion by blacksino** [CGAL/cgal#6403] [Q&A] [answered]
Hi, I am new to CGAL 3d mesh generation, I have generated multi-domain mesh by using labeled images. I am quite interested in those vertices and facets on boundaries of different domains, and I have no idea about how to retrieve those indexes of these vertices. I have tried using boost::get<int>(&(c3t3.index(vertex_handle)),however,it always return value 0.

Any suggestions?

Thanks.

**Comment by lrineau** [comment, marked as answer]
Your question is not precise enough. I guess you want to export the mesh to an index-based format, where CGAL meshes are pointer-based. The call `c3t3.index(vertex_handle)` will not help you, because that is the index of the sub-domain that contains the point of the vertex.

You need to compute an indexing of the triangulation. You can have a look at the I/O functions of CGAL Mesh_3, in particular the function `facets_in_complex_3_to_triangle_soup` in the file `include/CGAL/facets_in_complex_3_to_triangle_mesh.h`. You will see that the key of the indexing is a map of type `boost::unordered_map<Vertex_handle, std::size_t>` (as a complication, there is a special hash functor that is used to ensure determinism) that is used to index the vertices. And then the various traversal methods of the mesh are used.

You can also have a look at the header file `include/CGAL/internal/Mesh_3/Boundary_of_subdomain_of_complex_3_in_triangulation_3_to_off.h` that can export specifically the boundary of a given subdomain. Maybe that is what you want. Please do not use that header directly, because it is internal, and will be broken by next release of CGAL: copy-paste, and adapt, the code you need, instead.

**Reply by blacksino** [reply]
Thank you for your reply!
Yes, I do need to export the mesh to an index-based format, like Medit, indexed points, indexed faces, indexed tetrahedron(cells).
And my goal is to export the multi-domain mesh to vtk_file and label those points and faces on boundaries before export, then I  could distinguish those points and faces on domain boundaries in VTK easily.
 I do have checked the function you have mentioned above, I think it may be slightly different from what I would like to do. Because I would like to know the index relate to all points(facet, cell) in the tetra-mesh. Not the index relates to just on boundaries. 

I have tried to traverse all points and faces in meshing and determine whether it is on boundaries or not, however, I found it is difficult to do so without the `facets_in_complex` iterator because it can only iterate facets on the surface.

**Reply by blacksino** [reply]
A feasible method I can think of now is to create a hash table for the handles of all vertices in the process of traversing all cells in the mesh(using `Cells_in_complex_iterator`) and then traverse all the facets on the surface through an iterator such as `facets_in_complex`, If the facet is located on the surface of the subdomain, find the handles of the three vertices handle of the facet in the hash table, record their tags(they are located on the surface), and then export all the vertices and attribute records in the hash table together through the IO module in CGAL, is this approach feasible in CGAL?

**Reply by blacksino** [reply]
Oh, Got it,that's what exactly `output_to_tetgen `did.

**Reply by eudoxos** [reply]
Having the same issue, I could vertex pointers into the hash table (instead of hashing point coordinates), it works well:
```c++
// allocate cdt.number_of_vertices() and cdt.number_of_faces()
std::unordered_map<CDT::Vertex*,std::size_t> vert2ix;
for(auto const [i,vert]: std::views::enumerate(cdt.finite_vertex_handles())){
	vert2ix[&(*vert)]=i; // deref iterator, get pointer to it
	// save vert->point() into some array at [i]
}
for(auto const [i,face]: std::views::enumerate(cdt.finite_face_handles())){
	for(short j:{0,1,2}){
		CDT::Vertex* v=&(*(face->vertex(j)));
		assert(vert2ix.count(v)==1);
		size_t ix=vert2ix[v];
		// do something with the vertex index here
	}
}
```
